code/datums/helper_datums/construction_datum.dm

```diff
@@ -89,15 +119,94 @@
 		return FORWARD //to the first step -> forward
 	else if(L["backkey"] && istype(used_atom, L["backkey"]))
 		return BACKWARD //to the last step -> backwards
-	return 0
+	return FALSE
 
 /datum/construction/reversible/check_step(atom/used_atom,mob/user)
 	var/diff = is_right_key(used_atom)
 	if(diff)
 		if(custom_action(index, diff, used_atom, user))
 			update_index(diff)
-			return 1
-	return 0
+			return TRUE
+	return FALSE
 
 /datum/construction/reversible/custom_action(index, diff, used_atom, user)
-	return 1
+	return TRUE
+
+#define STATE_NEXT "next"
+#define STATE_PREV "prev"
+
+
+/datum/construction/reversible2
+	var/index
+	var/base_icon = "durand"
+
+/datum/construction/reversible2/New(atom)
+	..()
+	index = 1
+	return
+
+/datum/construction/reversible2/proc/update_index(diff as num, mob/user as mob)
+	index-=diff
+	if(index==steps.len+1)
+		spawn_result(user)
+	else
+		set_desc(index)
+	return
+
+/datum/construction/reversible2/proc/update_icon()
+	holder.icon_state="[base_icon]_[index]"
+
+/datum/construction/reversible2/is_right_key(mob/user as mob,atom/used_atom) // returns index step
+	var/list/state = steps[index]
+	if(STATE_NEXT in state)
+		var/list/step = state[STATE_NEXT]
+		if(istype(used_atom, step["key"]))
+			//if(L["consume"] && !try_consume(used_atom,L["consume"]))
+			//	return FALSE
+			return FORWARD //to the first step -> forward
+	else if(STATE_PREV in state)
+		var/list/step = state[STATE_PREV]
+		if(istype(used_atom, step["key"]))
+			//if(L["consume"] && !try_consume(used_atom,L["consume"]))
+			//	return FALSE
+			return BACKWARD //to the first step -> forward
+	return FALSE
+
+/datum/construction/reversible2/check_step(atom/used_atom,mob/user as mob)
+	var/diff = is_right_key(user,used_atom)
+	if(diff)
+		if(custom_action(index, diff, used_atom, user))
+			update_index(diff,user)
+			update_icon()
+			return TRUE
+	return FALSE
+
+/datum/construction/reversible2/proc/fixText(text,user)
+	text = replacetext(text,"{USER}","[user]")
+	text = replacetext(text,"{HOLDER}","[holder]")
+	return text
+
+/datum/construction/reversible2/custom_action(index, diff, used_atom, var/mob/user)
+	if(!..(index,used_atom,user))
+		return FALSE
+
+	var/list/step = steps[index]
+	var/list/state = step[diff==FORWARD ? STATE_NEXT : STATE_PREV]
+	user.visible_message(fixText(state["vis_msg"],user),fixText(state["self_msg"],user))
+
+	if("delete" in state)
+		qdel(used_atom)
+	else if("spawn" in state)
+		var/spawntype=state["spawn"]
+		var/atom/A = new spawntype(holder.loc)
+		if("amount" in state)

```

Comment:
you should not be checking for a key that has a value this way.

`if (state["spawn"] )` is much faster than `if("spawn" in state)` (but only works as long as state["spawn"] would normally return an actual value)